platforms/core-execution/persistent-cache/src/main/java/org/gradle/cache/internal/DefaultFileLockManager.java

```diff
@@ -286,6 +286,25 @@ public LockMode getMode() {
             return mode;
         }
 
+        /**
+         * Lock file is {@link java.io.RandomAccessFile} that has two regions:
+         * - lock state region, locked for the duration of the operation
+         * - lock info region, locked just to write the lock info or read info from it<br><br>
+         *
+         * We first try to acquire a lock on the state region with retries, see {@link #lockStateRegion(LockMode)}.
+         * If we use exclusive lock, and we succeed, we acquire a exclusive lock on the information region and write our details (port and lock id)
```

Comment:
❔ Why do we need to also lock the state region? If we already hold the lock for the info region, then nobody else will write to the state region. Acquiring a lock on the state region will also not prevent others from _reading_ the state region, right?

Reply:
"State region" lock is a "state" lock. We lock a directory state with it. This region is lock for the whole time of the process.

"Info region" lock is used just to exchange information between processes (but it's not used to actually lock a state of a directory). Important bit is, that only a process that has exclusive access to the state region writes to the info region. Another imporant bit is that info region is locked just for a short time when process needs to read/write to it. 

> Why do we need to also lock the state region?  If we already hold the lock for the info region, then nobody else will write to the state region.

Locking state region is more important than locking info region, since state region is actual lock that is used for the whole period of a process and locks the directory.
We need two separate regions, because if we use exclusive lock and just one region, then no other process will be able to read it since exclusive lock would forbid that. 

> Acquiring a lock on the state region will also not prevent others from reading the state region, right?

Here I think you meant "Acquiring a lock on the state region will also not prevent others from reading the ***info*** region". Yes I thinks so. After a process acquire state lock, there could be a small period of time till it acquires also info region

Reply:
I think it's fine how it is right now, we also link to the `lockStateRegion` in the docs. 

I think adding it to a class will make it more confusing, since there is also lots of different things going on there and this is specific to the lock method.